File: web/lib/types/supply.ts
// Sphere supply engine API contract.
// MIRRORS ai-services/app/models/supply.py (the canonical source). Any change to one
// must update the other in the SAME commit (see CLAUDE.md > Rules).

export type SupplyCategory = "water" | "food" | "shelter" | "nfi";

/** Inputs to the Sphere supply engine. */
export interface SupplyManifestRequest {
  /** People needing relief (e.g. TabPFN `affected`). >= 0. */
  predicted_affected: number;
  /** Provisioning horizon in days. >= 1. */
  days: number;
  /** Scales every supplied quantity for access constraints
   *  (1.0 = full; 0.5 = half deliverable; 1.2 = +20% buffer). >= 0. Defaults to 1.0. */
  access_modifier?: number;
  /** Caller key, echoed back on the manifest. */
  id?: string | null;
}

/** One supply line, fully traceable to the inputs that produced it. */
export interface SupplyLine {
  item: string;
  category: SupplyCategory;
  /** Unit of `quantity`, e.g. L, ration packs, blankets. */
  unit: string;
  /** Quantity to deliver (post access modifier). */
  quantity: number;
  /** Logistics weight per unit. */
  unit_weight_kg: number;
  /** quantity * unit_weight_kg (cargo demand for OR-Tools). */
  weight_kg: number;
  /** Human-readable formula behind `quantity`. */
  basis: string;
  /** Every multiplicand used (incl. access_modifier) for audit. */
  inputs: Record<string, number>;
}

/** A complete, auditable relief manifest for one affected population. */
export interface SupplyManifest {
  predicted_affected: number;
  days: number;
  access_modifier: number;
  /** ceil(affected / persons_per_household). */
  households: number;
  lines: SupplyLine[];
  /** Sum of line weights (cargo demand). */
  total_weight_kg: number;
  /** The Sphere/IFRC constants used, for audit. */
  standards: Record<string, number>;
  id?: string | null;
}

File: ai-services/app/models/supply.py
"""API contracts for the Sphere supply engine — SOURCE OF TRUTH (mirror in web/lib/types).

A manifest is an itemized, auditable relief order derived deterministically from a
predicted affected-population count. Every line records the exact inputs that produced
its quantity so a coordinator can trace (and override) any number. Consumed downstream by
OR-Tools routing (Phase 2.4), which is why each line also carries a logistics weight.
"""
from typing import Literal, Optional

from pydantic import BaseModel, Field


class SupplyManifestRequest(BaseModel):
    """Inputs to the Sphere supply engine."""

    predicted_affected: int = Field(
        ..., ge=0, description="people needing relief (e.g. TabPFN `affected`)"
    )
    days: int = Field(..., ge=1, description="provisioning horizon in days")
    access_modifier: float = Field(
        1.0, ge=0.0,
        description="scales every supplied quantity for access constraints "
                    "(1.0 = full; 0.5 = half deliverable; 1.2 = +20% buffer)",
    )
    id: Optional[str] = Field(None, description="caller key, echoed back on the manifest")


class SupplyLine(BaseModel):
    """One supply line, fully traceable to the inputs that produced it."""

    item: str = Field(..., description="human-readable item name")
    category: Literal["water", "food", "shelter", "nfi", "health"]
    unit: str = Field(..., description="unit of `quantity`, e.g. L, ration packs, blankets")
    quantity: int = Field(..., ge=0, description="quantity to deliver (post access modifier)")
    unit_weight_kg: float = Field(..., ge=0.0, description="logistics weight per unit")
    weight_kg: float = Field(..., ge=0.0, description="quantity * unit_weight_kg (cargo demand)")
    basis: str = Field(..., description="human-readable formula behind `quantity`")
    inputs: dict[str, float] = Field(
        ..., description="every multiplicand used (incl. access_modifier) for audit"
    )


class SupplyManifest(BaseModel):
    """A complete, auditable relief manifest for one affected population."""

    predicted_affected: int = Field(..., ge=0)
    days: int = Field(..., ge=1)
    access_modifier: float = Field(..., ge=0.0)
    households: int = Field(..., ge=0, description="ceil(affected / persons_per_household)")
    lines: list[SupplyLine] = Field(..., min_length=1)
    total_weight_kg: float = Field(..., ge=0.0, description="sum of line weights (cargo demand)")
    standards: dict[str, float] = Field(
        ..., description="the Sphere/IFRC constants used, for audit"
    )
    id: Optional[str] = None
